smarc_bt/src/mission_plan.py

```diff
@@ -473,6 +539,212 @@ def path_to_list(self, path_msg):
+            waypoints.append([(points_with_lolo[-1, 0], points_with_lolo[-1, 1])])
+            waypoints = [el for sublist in waypoints for el in sublist] # Flatten list of lists
+            waypoints_np = np.array(waypoints)
+            original_waypoints = waypoints_np.squeeze()  # Remove useless extra dimension
+
+        # Keep the other waypoints' parameters equal
+        mwp = mission.waypoints[0]
+        goal_tolerance = mwp.goal_tolerance
+        z_control_mode = mwp.z_control_mode
+        travel_altitude = mwp.travel_altitude
+        travel_depth = mwp.travel_depth
+        speed_control_mode = mwp.speed_control_mode
+        travel_rpm = mwp.travel_rpm
+        travel_speed = mwp.travel_speed
+
+        # Compute angle between waypoints and get the new wps array
+        waypoints_complete, _ = waypoints_with_yaw(original_waypoints)
+
+        path = []
+        for j in range(len(waypoints_complete)-1):
+            param = calc_dubins_path(waypoints_complete[j], waypoints_complete[j+1], turn_radius)
+            path.append(dubins_traj(param,1))
+
+        dubins_waypoints = []
+        # Only define the new waypoints on the original waypoint and on the curve
+        # Each element of path is an array of points between on waypoint and the next one
+        for i, el in enumerate(path):
+            # Compute the differnece between orientation 
+            # The maxima represent the points on the curve
+            delta_angles = abs(np.diff(el[:, 2]))
+            delta_angles = [0] + delta_angles
+            max_idxs = np.argpartition(delta_angles, -num_points)[-num_points:] # Keep two points in each curve
+            max_idxs = np.sort(max_idxs)
+            for ind in max_idxs:
+                dubins_waypoints.append(el[ind]) # A row of path[i] represents a single point
+
+        # Sort the waypoints
+        full_path = []
+        ordered_idxs = []
+        for el in path:
+            for e in el:
+                full_path.append([e[0], e[1], e[2]])
+        for j, el in enumerate(dubins_waypoints):
+            try:
+                ordered_idxs.append(full_path.index([el[0], el[1], el[2]])) # Get the dubins wp index in the full path
+            except ValueError:
+                rospy.logwarn("Point " + str(j) + " not found in the path!")
+        ordered_idxs = np.sort(ordered_idxs)
+        dubins_waypoints_ordered = [full_path[i] for i in ordered_idxs]
+
+        # Include original waypoints
+        # The first one is lolo's position, we don't need it
+        wp_i = num_points
+        for wp in waypoints_complete[1:-1]:
+            dubins_waypoints_ordered.insert(wp_i, [wp.x, wp.y, wp.psi])
+            wp_i += num_points + 1
+        dubins_waypoints_ordered.append([waypoints_complete[-1].x, waypoints_complete[-1].y, waypoints_complete[-1].psi])
+        dubins_waypoints_final = dubins_waypoints_ordered[1:]
+
+        dubins_mission = MissionControl()
+        dubins_mission = mission
```

Comment:
over-write?

Reply:
Personally I prefer having them as two separate objects, but I could definitely directly edit the original mission

Reply:
but if you're gonna over-write the `dubins_mission` object with the `mission` one, why even construct the `MissionControl` object right before?

Reply:
Fair enough